server/src/main/java/org/opensearch/index/mapper/DateFieldMapper.java

```diff
@@ -226,6 +232,58 @@ private static DateFieldMapper toType(FieldMapper in) {
         return (DateFieldMapper) in;
     }
 
+    @Override
+    protected void deriveSource(XContentBuilder builder, LeafReader leafReader, int docId) throws IOException {
+        possibleToDeriveSource();
+        boolean isStoredFieldPresent = false;
+        // 1. Lookup stored field, which will help in preserving order of values in case of multi value field
+        // 2. If field value is not found using stored field, lookup doc values
+        if (mappedFieldType.isStored()) {
+            List<Object> values = new ArrayList<>();
+            SingleFieldsVisitor singleFieldsVisitor = new SingleFieldsVisitor(mappedFieldType, values);
+            StoredFields storedFields = leafReader.storedFields();
+            storedFields.document(docId, singleFieldsVisitor);
+            if (!values.isEmpty()) {
+                if (values.size() > 1) {
+                    builder.array(name(), values.toArray());
+                } else {
+                    builder.field(name(), values.getFirst());
+                }
+                isStoredFieldPresent = true;
+            }
+        }
+        if (isStoredFieldPresent) {
+            return;
+        }
+        SortedNumericDocValues sortedNumericDocValues = leafReader.getSortedNumericDocValues(name());
+        if (sortedNumericDocValues != null && sortedNumericDocValues.advanceExact(docId)) {
+            int valueCount = sortedNumericDocValues.docValueCount();
+            if (valueCount > 0) {
+                String[] values = new String[valueCount];
+                DateFormatter dateFormatter = fieldType().dateTimeFormatter;
+                for (int i = 0; i < valueCount; i++) {
+                    values[i] = dateFormatter.formatMillis(sortedNumericDocValues.nextValue());
```

Comment:
Can you add documentation stating that this is not guaranteed to be same as what was provided in the input?